File: src/main/java/org/jcodec/containers/mkv/boxes/EbmlUint.java
package org.jcodec.containers.mkv.boxes;

import java.nio.ByteBuffer;

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed under FreeBSD License
 * 
 * EBML IO implementation
 * 
 * @author Jay Codec
 * 
 */
public class EbmlUint extends EbmlBin {

    public EbmlUint(byte[] id) {
        super(id);
    }
    
    public EbmlUint(byte[] id, long value) {
        super(id);
        set(value);
    }
    
    public void set(long value){
        this.data = ByteBuffer.wrap(longToBytes(value));
        this.dataLen = this.data.limit();
    }

    public long get() {
        long l = 0;
        long tmp = 0;
        for (int i = 0; i < data.limit(); i++) {
            tmp = ((long) data.get(data.limit() - 1 - i)) << 56;
            tmp >>>= (56 - (i * 8));
            l |= tmp;
        }
        return l;
    }
    
    public static byte[] longToBytes(long value) {
        byte[] b = new byte[calculatePayloadSize(value)];
        for (int i = b.length - 1; i >= 0; i--) {
            b[i] = (byte) (value >>> (8 * (b.length - i - 1)));
        }
        return b;
    }

    public static int calculatePayloadSize(long value) {
        if (value == 0)
            return 1;
        
        long mask = 0xFF00000000000000L;
        int i = 0;
        while ((value & (mask>>>8*i)) == 0 && i < 8)
            i++;
        
        return 8 - i;
    }

}

File: src/main/java/org/jcodec/containers/mkv/boxes/EbmlUint.kt
package org.jcodec.containers.mkv.boxes

import java.nio.ByteBuffer

/**
 * This class is part of JCodec ( www.jcodec.org ) This software is distributed under FreeBSD License
 *
 * EBML IO implementation
 *
 * @author The JCodec project
 */
class EbmlUint(id: ByteArray) : EbmlBin(id) {

    var uint: Long
        get() {
            var l: Long = 0
            var tmp: Long = 0
            val d = _data!!
            for (i in 0 until d.limit()) {
                tmp = d[d.limit() - 1 - i].toLong() shl 56
                tmp = tmp ushr 56 - i * 8
                l = l or tmp
            }
            return l
        }
        set(value) {
            _data = ByteBuffer.wrap(longToBytes(value))
            _dataLen = _data!!.limit()
        }

    companion object {
        @JvmStatic
        fun createEbmlUint(id: ByteArray, value: Long): EbmlUint {
            val e = EbmlUint(id)
            e.uint = value
            return e
        }

        @JvmStatic
        fun longToBytes(value: Long): ByteArray {
            val b = ByteArray(calculatePayloadSize(value))
            for (i in b.indices.reversed()) {
                b[i] = (value ushr 8 * (b.size - i - 1)).toByte()
            }
            return b
        }

        @JvmStatic
        fun calculatePayloadSize(value: Long): Int {
            if (value == 0L) return 1
            return if (value <= 0x7fffffffL) {
                4 - (Integer.numberOfLeadingZeros(value.toInt()) shr 3)
            } else 8 - (java.lang.Long.numberOfLeadingZeros(value) shr 3)
        }
    }
}